plugins/inventory/hrobot.py

```diff
@@ -0,0 +1,180 @@
+            required: true
+            env:
+                - name: HROBOT_API_PASSWORD
+        filters:
+          description:
+              - A dictionary of filter value pairs.
+              - Available filters are listed here are keys of server
+          type: dict
+          default: {}
+"""
+
+EXAMPLES = r"""
+# Fetch all hosts in Hetzner Robot
+plugin: hrobot
+# Filters all servers in ready state
+filters:
+  status: ready
+# Example using constructed features to create groups
+plugin: hrobot
+filters:
+  status: ready
+  traffic: unlimited
+# keyed_groups may be used to create custom groups
+strict: False
+keyed_groups:
+  # Add e.g. dc hosts
+  - key: dc
+    separator: ""
+"""
+
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+from ansible.module_utils.urls import open_url
+from ansible.utils.display import Display
+from ansible.errors import AnsibleError
+
+import json
+
+from ansible_collections.community.hrobot.plugins.module_utils.robot import (
+    BASE_URL
+)
+
+display = Display()
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+
+    NAME = 'hrobot'
+
+    def verify_file(self, path):
+        ''' return true/false if this is possibly a valid file for this plugin to consume '''
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            # base class verifies that file exists and is readable by current user
+            if path.endswith(('hrobot.yaml', 'hrobot.yml')):
+                valid = True
+        display.debug("hrobot inventory filename must end with 'hrobot.yml' or 'hrobot.yaml'")
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        super(InventoryModule, self).parse(inventory, loader, path)
+        servers = {}
+        config = self._read_config_data(path)
+        self.load_cache_plugin()
+        cache_key = self.get_cache_key(path)
+
+        # cache may be True or False at this point to indicate if the inventory is being refreshed
+        # get the user's cache option too to see if we should save the cache if it is changing
+        user_cache_setting = self.get_option('cache')
+
+        # read if the user has caching enabled and the cache isn't being refreshed
+        attempt_to_read_cache = user_cache_setting and cache
+        # update if the user has caching enabled and the cache is being refreshed; update this value to True if the cache has expired below
+        cache_needs_update = user_cache_setting and not cache
+
+        # attempt to read the cache if inventory isn't being refreshed and the user has caching enabled
+        if attempt_to_read_cache:
+            try:
+                servers = self._cache[cache_key]
+            except KeyError:
+                # This occurs if the cache_key is not in the cache or if the cache_key expired, so the cache needs to be updated
+                cache_needs_update = True
+        else:
```

Comment:
```suggestion
        elif not cache_needs_update:
```
To avoid `get_servers()` being called twice.